### ISSUE TITLE
[vue3] When is 8.0 releasing?

### ISSUE THREAD

**Issue by ninjabhishek** [vuejs/vue-class-component#584]
It looks like v8 rc was released a long time ago. When is the final version expected?

**Comment by ninjabhishek** [comment]
Agreed. Vue 3 typescript class based project works with 8.0, and working on rc candidate is a security concern in a corporate environment. They should release or someone(or me) will have to take a fork and release it with different name and proper versioning.

**Comment by ruojianll** [comment]
This work is very important for VUE 3.

**Comment by jaeming** [comment]
Agree as well. This is the major blocker for my team in updating a very large codebase to Vue 3. I remember hearing assurances from Vue core devs that the class syntax would not be abandoned in favor of composition api but I'm starting to wonder if this is the case anymore.

**Comment by ghiscoding** [comment]
Side note which might help pushing this, [Vue 3 is now default](https://blog.vuejs.org/posts/vue-3-as-the-new-default.html) aka Vue 3 is now released has of Feb 7 🚀

**Comment by donggyu04** [comment]
@ktsn This feature is very important to me. Do you have any progress or update?

**Comment by andyfu6** [comment]
@ktsn Why not provide Vue constructor like before?

**Comment by nseb** [comment]
any replies ?

**Comment by pkzOR** [comment]
I'm looking for status updata of v8 and wondering about it's long term viability. Can anyone provide some insight?

**Comment by pkzOR** [comment]
> I'm looking for status update of v8 and wondering about it's long term viability. Can anyone provide some insight?

To answer my own question, it appears that Vue no longer recommends class-components (https://vuejs.org/guide/extras/composition-api-faq.html#relationship-with-class-api)  and so there is little incentive to maintain this.

**Comment by nseb** [comment]
@pkzOR  , Thanks , I just read.
About me , I hate the composition setup that I find it makes maintenance difficult because the code is unreadable. I come from the Java world, I use TS and VueClassComponent on a big project, I find the top balance between ease of development and maintenance.

**Comment by ninjabhishek** [comment]
Vue 3 ecosystem is messed up right now. The core libraries published by vue is interdependent, and so if one is not released, the others, even if released carry unsecured versions. One example would be https://github.com/vuejs/test-utils/blob/main/package.json where it is an rc version carrying the rc of vue-class-components. And then vue-cli-jest v5 carries v27-rc of vue jest. Vue 3 is not ready for enterprise yet. One scan, and we are under red alert

**Comment by ImBoop** [comment]
> > I'm looking for status update of v8 and wondering about it's long term viability. Can anyone provide some insight?
> 
> To answer my own question, it appears that Vue no longer recommends class-components (https://vuejs.org/guide/extras/composition-api-faq.html#relationship-with-class-api) and so there is little incentive to maintain this.

That's actually nuts. Their reasoning isn't sound at all, class API is more in line with the direction the web is moving, and is also far easier for developers to use. Angular, Web Components, and even React have first class (no pun intended) support for class based components. Vue randomly deprecating/removing features and having no sound strategy make it hard to justify using it in the future.

**Comment by WolfspiritM** [comment]
> That's actually nuts. Their reasoning isn't sound at all, class API is more in line with the direction the web is moving, and is also far easier for developers to use. Angular, Web Components, and even React have first class (no pun intended) support for class based components. Vue randomly deprecating/removing features and having no sound strategy make it hard to justify using it in the future.

Yep. We also moved away from Vue to Angular and are currently rebuilding our components in Angular cause of that. It's sad but I also don't recommend using Vue anymore cause of decisions like that.

**Comment by BuddhiAbeyratneAL** [comment]
Im kinda ashamed to say this but as a TL this was a rookie mistake trusting in a org this young. We'll be moving to angular too for our new projects (hate my gut for this as I love class based vue and Vuex combo that Nuxt provided) 
it feels like the lac of good videos / docs and articles that was pushing for OOP standards made the path a not so well trodden one making it easier for replacement. 

I still have a bit of hope that they will go to the http://typescript.nuxtjs.org/cookbook/components page and click the three options and see what's more readable and at least put some maintenance into it. 

In the meantime our org decided to take our chances with going for micro front ends. we'll be breaking what we have and freezing some parts of the application and will move into other frameworks for the time being. 

If any of you have code mods or migrations paths please let me know.
Thinking of starting a codemod repo to do this.

**Comment by PashaTurok** [comment]
The same problem. I read about Vue and liked to its simplicity. However, as a professional developer I use only OOP. I started a project with Vue3 having great hopes on `vue-class-component`. However, no documentation for v. 8, project hasn't been updated for a long time, no any answer from @ktsn . Unfortunately to start something serious in such conditions is too risky. So, goodbye, Vue and hello, Angular.

PS. No, really, it is 21'st century. Why wasn't Vue3 developed using OOP?

**Comment by Fleeck** [comment]
My two cents after a few months: Composition API has completely replaced `vue-class-component` for me. Try it out.

**Comment by WolfspiritM** [comment]
> My two cents after a few months: Composition API has completely replaced `vue-class-component` for me. Try it out.

The problem is that with a huge codebase there is no migration strategy between vue-class-component and composition api. It's a tough decision to rewrite everything to composition API without knowing if they decide to abondon that in the future again out of some reason or rather use a more "developer/large codebase friendly" solution like Angular where a migration to a new version is automated as good as possible.

**Comment by chubbard** [comment]
> It's funny because Vue gained a lot of popularity due to the AngularJS to Angular 2 era was even dubbed by a lot of fans as the "real" angular 2. Now it seems to be making the same mistakes.

I had picked Vuejs for it's simplicity and closer relation to AngularJS after getting burned by AngularJS being effectively dropped.  Now I'm twice burned.  I've read quite a bit about the composition API, and I can't see anything that demonstrates it's superior architecture.  Sure the make claims left and right about how awesome, but nothing I've read convinces me it's any better than class-based.  There is a significant amount of boiler plate code composition API creates that didn't exist with class-based API.  It's only "better" as a matter of taste not an objective truth.

**Comment by BuddhiAbeyratne** [comment]
> My two cents after a few months: Composition API has completely replaced `vue-class-component` for me. Try it out.

I mean you can see that the people crying over here are Leeds having to migrate everything they worked on for 4 years into a  completely new framework and workflow and standards and dev tools and libs. By the time you are done with all of this if you are an SME may god be with you while looking at a recession and cost cuts. If you are a smaller org GG WP better luck next time. If you were an agency say bye to all the clients you recommended this to when you ask them for budget increases to mitigate to prevent catastrophic failure via deprecation. 
I don't think you understand the severity of this entire fiasco. Im flabbergasted that more YouTubers and influencers aren't talking about this. Its the nitty gritty stuff that might not get all the views but damn I wish I had the platform to prevent someone just like me who 3 years back saw vue2 as a great option to not  fall into the same pitfall again with Vue3. 

The maintainers have forgotten that what makes or breaks a framework is adoption. Its people. Angular had to fight its way out of that mess and still people are sceptical about it because of the js to angular 2 fiasco. Its ratings are low and adoption is low due to the same reason. 

What do you think will happen when people understand how flaky the community is. You mean to tell me no one in the core team can see this thread ? Why isn't there a formal response ? at least Nuxt is working on bridge. but that too for how long when its basis is Vue and SME's and Enterprise starts fleeing from it. when startups with good CTO's and leads start avoiding it. 

They made promises that this will be maintained. what happen to all of that absolute garbage RFC promises.

**Comment by ImBoop** [comment]
I've considered forking Vue 3, porting some of the nice to haves from Vue 2 that 3 removed, and having class based be a first class citizen.

The problem, of course, is finding the time, and yet another framework just fragments an already small user base, so getting support is optimistic.

**Comment by chubbard** [comment]
The argument from Vue was that they didn't want to support 3 different ways to do the "same thing": Options, Composition, Classes which is the only substantive argument put forth from reading what little was said about it.  The FAQ doesn't even try to give an argument and just unceremoniously denounces it without any justification.   Quite telling about the attitude on the team towards this idea.  However, as pointed out in this thoughtful post: 

https://github.com/vuejs/rfcs/discussions/276

There wouldn't need to be large ceremony to give 1st class sanctioning of a different paradigm.  It's mostly just recognizing it is a way of working with Vue and providing docs.  However, when this is raised it's met with dismissive resistance and "Yeah but" arguments which shows that the community isn't open to this idea no matter its merits.  So if this is a hill you gotta die on you'll have to reconstruct a new hill, pick up the pieces, and recreates typescript class support in a new project.  I thought of the same thing, but then again who has the time?

**Comment by ImBoop** [comment]
> The argument from Vue was that they didn't want to support 3 different ways to do the "same thing": Options, Composition, Classes which is the only substantive argument put forth from reading what little was said about it. The FAQ doesn't even try to give an argument and just unceremoniously denounces it without any justification. Quite telling about the attitude on the team towards this idea. However, as pointed out in this thoughtful post:
> 
> [vuejs/rfcs#276](https://github.com/vuejs/rfcs/discussions/276)
> 
> There wouldn't need to be large ceremony to give 1st class sanctioning of a different paradigm. It's mostly just recognizing it is a way of working with Vue and providing docs. However, when this is raised it's met with dismissive resistance and "Yeah but" arguments which shows that the community isn't open to this idea no matter its merits. So if this is a hill you gotta die on you'll have to reconstruct a new hill, pick up the pieces, and recreates typescript class support in a new project. I thought of the same thing, but then again who has the time?

That's wild considering they don't really want people using options over composition, so it seems to me they should have deprecated options instead and kept class-based for the 2 systems with first class support.

**Comment by Fleeck** [comment]
> > My two cents after a few months: Composition API has completely replaced `vue-class-component` for me. Try it out.
> 
> I mean you can see that the people crying over here are Leeds having to migrate everything they worked on for 4 years into a completely new framework and workflow and standards and dev tools and libs. By the time you are done with all of this if you are an SME may god be with you while looking at a recession and cost cuts. If you are a smaller org GG WP better luck next time. If you were an agency say bye to all the clients you recommended this to when you ask them for budget increases to mitigate to prevent catastrophic failure via deprecation. I don't think you understand the severity of this entire fiasco. Im flabbergasted that more YouTubers and influencers aren't talking about this. Its the nitty gritty stuff that might not get all the views but damn I wish I had the platform to prevent someone just like me who 3 years back saw vue2 as a great option to not fall into the same pitfall again with Vue3.
> 
> The maintainers have forgotten that what makes or breaks a framework is adoption. Its people. Angular had to fight its way out of that mess and still people are sceptical about it because of the js to angular 2 fiasco. Its ratings are low and adoption is low due to the same reason.
> 
> What do you think will happen when people understand how flaky the community is. You mean to tell me no one in the core team can see this thread ? Why isn't there a formal response ? at least Nuxt is working on bridge. but that too for how long when its basis is Vue and SME's and Enterprise starts fleeing from it. when startups with good CTO's and leads start avoiding it.
> 
> They made promises that this will be maintained. what happen to all of that absolute garbage RFC promises.

I totally understand your frustration and I'd love to hear an official statement on this. As a side note it was quite tedious for me to rewrite everything to Composition API and I wish we had at least a migration strategy.

**Comment by jaeming** [comment]
It's probably not worth discussing anymore here unless one of us wants to fork or propose to maintainers someone who is willing to take over. For those looking to migrate, something as naive as this will work in some simple cases:

![class ref](https://user-images.githubusercontent.com/2175262/163862876-a32db60d-3496-4164-ba68-cb9803ddfacf.png)

although you will probably want to go ahead and do a full refactor to avoid a bunch of issues:

![refactor](https://user-images.githubusercontent.com/2175262/163864100-160d246b-5d9d-41d0-9cda-b21f432aae44.png)

I know a lot of people would say this is not that big of a refactor...but when you have a huge project (14,000+ commits, with 40+ devs contributing), that's a lot of re-training, refactoring, and QA regression testing you have to justify and then explain to management why you should continue on the Vue path when everyone else in the industry to telling them to React or Angular are the goto solutions.

**Comment by chubbard** [comment]
If this project was finished and released it wouldn't really matter what Vue team did or said.  There would be a viable project supporting class based Vue.  This project was being worked on then abruptly stopped, but every time people asked about class based support in Vue 3 they said it was "Done", but the documentation just needed to be written.  Then it never got written so is all that is left to make this work is write the docs?  If not what else has to be done to release the next branch?

**Comment by ruojianll** [comment]
@chubbard Agree. In my opinion, the document should be written in class based components instead of current version. Class based components, jsx and TypeScript are important for modern javascript codes to build a large complex project. The document has been written in an embarrassed way.

**Comment by ruojianll** [comment]
@jaeming `<script setup>` is more magical and only used by vue. Vue single file component is also magical. [Decorator pattern has been common used by many program languages as `@` syntax, class or function](https://en.wikipedia.org/wiki/Decorator_pattern#:~:text=In%20object%2Doriented%20programming%2C%20the,objects%20from%20the%20same%20class.). So in vue2 I use this repo with typescript  tsx and class based component, and divide them in to 3 files(class, render function, css) to build my component and import them. SFC is not easy to be maintained in a large component.

**Comment by BuddhiAbeyratneAL** [comment]
@ruojianll exact way I'm using react these days. Team managed to pull a miracle and migrate the stuff out to react in record time. thanks to our components being broken down well and having separation of biz logic and api integration and store from the components. We were using vue only to make UIs work and to bind a bit of data. Highly recommended if you have something like this already. the composition mess can be sorted by using prop.children properly and using the context hook for micro states for small things. We using mobx for state management its just like vue and even has deep observable and out of the box gave 20% better performance from our painstakingly designed Vuex store. we just thew things to gather and it stuck when we did a spike to test things out. Best week of figuring shit out I've had in a while. they have improved massively and if you use nextjs mobx and functional programming only for the UI life becomes a breeze. best thing is its not as wishy washy set of fever dream experiments driven by a stone walling toxic ex of a core team. https://github.com/infinitered/ignite 
This is made for react native but it helps understand better design rational. 
NextJS is a solid choice as a Nuxt alternative. I'll report back after the honeymoon phase too. 
If possible I'll make an automation tool to help people port Nuxt projects to next.

**Comment by ruojianll** [comment]
I set up a repo to do this work in vue3.
https://github.com/facing-dev/vue-facing-decorator

**Comment by ruojianll** [comment]
first release 
https://www.npmjs.com/package/vue-facing-decorator

**Comment by ruojianll** [comment]
> I set up a repo to do this work in vue3. https://github.com/facing-dev/vue-facing-decorator

First release.
https://www.npmjs.com/package/vue-facing-decorator

**Comment by zhoulujun** [comment]
This work is very important for VUE 3. and me

**Comment by lzxb** [comment]
> It's probably not worth discussing anymore here unless one of us wants to fork or propose to maintainers someone who is willing to take over. For those looking to migrate, something as naive as this will work in some simple cases:
> 
> ![class ref](https://user-images.githubusercontent.com/2175262/163862876-a32db60d-3496-4164-ba68-cb9803ddfacf.png)
> 
> although you will probably want to go ahead and do a full refactor to avoid a bunch of issues:
> 
> ![refactor](https://user-images.githubusercontent.com/2175262/163864100-160d246b-5d9d-41d0-9cda-b21f432aae44.png)
> 
> I know a lot of people would say this is not that big of a refactor...but when you have a huge project (14,000+ commits, with 40+ devs contributing), that's a lot of re-training, refactoring, and QA regression testing you have to justify and then explain to management why you should continue on the Vue path when everyone else in the industry is telling them to React or Angular are the goto solutions.

![image](https://user-images.githubusercontent.com/8424643/186557011-c76ea02e-458c-4060-a97d-f814cceaf16b.png)


I created a repository to solve this problem, As an alternative to vue-class-component, we will use it in large-scale projects

https://github.com/fmfe/vue-class-setup

**Comment by snake-py** [comment]
So, I am in the process of migrating 3 large code bases to vue 3 is there a good solution to support class-based-components in Vue 3? At least for the time, I am rewriting every component.

**Comment by viT-1** [comment]
> I set up a repo to do this work in vue3. https://github.com/facing-dev/vue-facing-decorator

My [old setup](https://github.com/viT-1/systemjs-ts-es6-vue) with SystemJS (as a bundler) support, without SFC and with ts-jest.

**Comment by rdhelms** [comment]
I think @ruojianll 's vue-facing-decorator has a ton of potential...of all the alternatives to the original vue-class-component and vue-property-decorator configs that I've seen so far, that seems to work the best with our existing code with minimal changes. Has anyone else had a chance to check that out? I would love if we could come up with an official vue3 successor to these packages, since the maintainers here don't seem to be responding anymore

https://facing-dev.github.io/vue-facing-decorator/
https://github.com/vuejs/vue-class-component/issues/569#issuecomment-1145882492

**Comment by akoidan** [comment]
I'm using this setup for vue3 - vue-class-component
https://github.com/akoidan/vue3-vite-vue-class-component

**Comment by snake-py** [comment]
We decided that it is too much of an effort and are currently in the process of throwing vue out. Our managers do not feel comfortable proceeding with vue, if this is the behavior of the vue team once they switch their opinion. I know this lib was not "official", but it was officially recommended and Even You's account hade much impact on this lib. SO it feels really awful that it just got abandoned.

**Comment by jaeming** [comment]
We also made the decision to leave Vue. It's kind of funny that they garnered so much market share from angular.js users who were disenchanted with the changes and upgrade path to angular 2+ and now they follow in their footsteps by making the same mistake.

**Comment by ingram90** [comment]
We just completed migrating a 300+ components project from Vue 2 to Vue 3 with the help of [vue-facing-decorator](https://facing-dev.github.io/vue-facing-decorator/).

There are a few minor glitches that need to be tweaked, but it's not a big deal.

I highly recommend vue-facing-decorator for all of us who are tackling the migration to Vue 3.

Thanks to the work of @ruojianll.

**Comment by nseb** [comment]
@ingram90 
thank you for your return. What are the small problems encountered or were they specific to your project?

**Comment by rdhelms** [comment]
@nseb We're also in the process of migrating several apps with 100+ components to Vue 3 using vue-facing-decorator, and so far almost all of the work has been from the official list of Vue 3 breaking changes, which would be issues no matter how you migrate the components themselves. The component "migration" was just doing a global find-and-replace of vue-property-decorator or vue-class-component with vue-facing-decorator.

**Comment by rdhelms** [comment]
Also, vue-facing-decorator is officially linked as a suggested package in the vue-class-component readme and docs now, for anyone who hadn't seen that.

**Comment by BuddhiAbeyratneAL** [comment]
@rdhelms Id love to get to know more about that journey, a blog when you have time maybe ?

**Comment by nseb** [comment]
@rdhelms 
Thank you for your feedback, it's interesting, I have the same number of components but I'm in vue3 with vue-class-component 8.0-rc1, I wanted to switch to vue-facing-decorator

**Comment by ingram90** [comment]
@nseb 

Most of our issues may be related to [vue-facing-decorator](https://facing-dev.github.io/vue-facing-decorator/), but I am not sure:

* We have replaced `@Emit()` with `this.$emit()` because there are some cases where it does not work. We have not isolated the problem yet, so we have changed everything to `this.$emit()` for now.

* Static members in class components are not working, but this is an easy fix.

* We have encountered something strange like this:

```
@Provide()
@Prop()
foo!: string;
```

This does not work with facing decorator, but I don't think anyone writes it this way :)

* We are not using mixins, so I am not sure if they are reliable or not.

**Comment by rdhelms** [comment]
@ingram90 @nseb We also changed our `$emit` calls to `@Emit`, but that's primarily due to the [Vue 3 breaking change](https://v3-migration.vuejs.org/breaking-changes/emits-option.html) that emitted component events need to be declared with the `emits` option. The alternative is to just declare the events in the `emits` option per the docs, but using the v-f-d `@Emit` decorator does that for you.

**Comment by nseb** [comment]
@rdhelms thx
 How migrate multi inheritance 
 in vue-class-component : 

 HomeLayout extends mixins(UMBaseComponent, AuthComponent) {
 
 in vue facing decorator
@Componenet({
mixins : [AuthComponent]
})
 HomeLayout extends extends UMBaseComponent {
 
 it seems to be on the market, but my IDE (IntellijIdea) does not see the methods inherited from AuthComponent and puts me in error , I did not have the problem with vue-class-component

**Comment by ruojianll** [comment]
@nseb I add new feature, which is similar to vue-class-component's mixins function to extends multiple components and keep type details .
https://facing-dev.github.io/vue-facing-decorator/#/en/inheritance/component/component?id=extends-multiple-components
v2.1.19

**Comment by BuddhiAbeyratneAL** [comment]
Has anyone used it with nuxt 3 ?

**Comment by nseb** [comment]
@ruojianll 
Excellent, top rap, I'm testing this right away.
I had other components in vu-class-component with computed mapGetters , I can use it or I have to migrate?

@Options({
  computed: {
    ...mapGetters('events', {
      currentEvent: 'getCurrentEvent'
    })
  }

**Comment by ruojianll** [comment]
@nseb  You should do something to migrate.

**Comment by BuddhiAbeyratneAL** [comment]
@ruojianll My faith in the community is slowly being restored thanks to you 
How can we help you ?

**Comment by rdhelms** [comment]
@BuddhiAbeyratne There's a new vue-facing-decorator beta version that @ruojianll just released which targets support for stage 3 decorators - it would be great for everyone thinking of using v-f-d to provide feedback there. And @ruojianll has been very responsive in issues and discussions there, as well as in the v-f-d discord server.
https://github.com/facing-dev/vue-facing-decorator/discussions/44

**Comment by ruojianll** [comment]
@BuddhiAbeyratne 
As @rdhelms said, a [beta version](https://facing-dev.github.io/vue-facing-decorator/#/en/next-version/next-version) for stage 3 decorators released. You could try it and feedback in here https://github.com/facing-dev/vue-facing-decorator/discussions/44
You could also PR the features what you want and help us to improve out document translation in enlish.
https://facing-dev.github.io/vue-facing-decorator/#/?id=document-translations-wanted
Thank you.

**Comment by Haixing-Hu** [comment]
Hello everyone,

I've created a brand-new vue-class-component for Vue 3. It's implemented in pure JavaScript, eliminating the need for TypeScript configuration. Additionally, it incorporates the most recent (as of May 2023) [stage 3 proposal of JavaScript decorators](https://github.com/tc39/proposal-decorators) and [stage 3 proposal of JavaScript decorator metadata](https://github.com/tc39/proposal-decorator-metadata).

You can find it here: [vue3-class-component](https://github.com/Haixing-Hu/vue3-class-component)

We welcome feedback, bug reports, and contributions. Feel free to get involved!

**Comment by nseb** [comment]
What's is the difference with vue-facind-decorator?

**Comment by Haixing-Hu** [comment]
@nseb 

1. The code is written in pure JavaScript, not TypeScript, so there is no need to configure TypeScript for your Vue project. Personally, I prefer the simplicity of JavaScript's compared to the complex type system of TypeScript.
3. It supports the stage 3 proposal of JavaScript decorators.
4. It has been thoroughly tested and has achieved 100% code coverage.
5. The source code is written in a clean and simple manner, making it easy for anyone to understand.

**Comment by zhennann** [comment]
Hello everyone, I have created a vue3 framework with ioc container, named as `Cabloy-Front`.

You can find it here: [https://github.com/cabloy/cabloy-front](https://github.com/cabloy/cabloy-front)

Do you agree with such a concept: Class should not be used in the `view layer`, but in the `business layer`.

Therefore, it is completely different from `vue-facind-decorator` or `vue-class-component`  that the Vue3 setup syntax sugar is still used to define vue components (such as props and emits of components), and `ioc container` and `class` are introduced only at the business level. In this way, combine the advantages of the two (function/class) to make the code more concise and elegant.

Cabloy-Front has a very important feature: With the support of ioc container, defining reactive states no longer needs `ref/reactive`. Without `ref`, naturally there is no need to write a lot of `ref.value`.

## Demonstration: no `ref/reactive`, no `ref.value`

[![Edit zhennann/cabloy-front-demo-codesandbox/main](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/github/zhennann/cabloy-front-demo-codesandbox2/main?embed=1&file=%2Fsrc%2Fsuite%2Fa-demo%2Fmodules%2Fa-demo%2Fsrc%2Fpage%2Fcounter%2Fcontroller.ts)

### 1. Define reactive state

Define a reactive variable `count` in the component and add two methods to modify its value

```typescript
export class MotherPageCounter {
  count: number = 0;

  inrement() {
    this.count++;
  }

  decrement() {
    this.count--;
  }
}
```

### 2. Use reactive state

Use `count` in render class

```typescript
export class RenderPageCounter {
  render() {
    return (
      <div>
        <div>count(ref): {this.count}</div>
        <button onClick={() => this.inrement()}>Inrement</button>
        <button onClick={() => this.decrement()}>Decrement</button>
      </div>
    );
  }
}
```

## Demonstration: dependency injection

[![Edit zhennann/cabloy-front-demo-codesandbox/main](https://codesandbox.io/static/img/play-codesandbox.svg)](https://codesandbox.io/p/github/zhennann/cabloy-front-demo-codesandbox2/main?embed=1&file=%2Fsrc%2Fsuite%2Fa-demo%2Fmodules%2Fa-demo%2Fsrc%2Fpage%2Fcounter2%2Fcontroller.ts)

### 1. Logic Reuse

Create a `Counter` Bean to implement the logic of `count`

```typescript
@Local()
export class Counter {
  count: number = 0;

  inrement() {
    this.count++;
  }

  decrement() {
    this.count--;
  }
}
```

### 2. Inject and use in a component

```typescript
export class MotherPageCounter {
  @Use()
  $$counter: Counter;
}
```

```typescript
export class RenderPageCounter {
  render() {
    return (
      <div>
        <div>count(ref): {this.$$counter.count}</div>
        <button onClick={() => this.$$counter.inrement()}>Inrement</button>
        <button onClick={() => this.$$counter.decrement()}>Decrement</button>
      </div>
    );
  }
}
```
